src/core/ext/transport/chttp2/transport/hpack_table.h

```diff
@@ -46,41 +47,66 @@
 #endif
 
 /* hpack decoder table */
-typedef struct {
+struct grpc_chttp2_hptbl {
+  static uint32_t entries_for_bytes(uint32_t bytes) {
+    return (bytes + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+           GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+  }
+  static constexpr uint32_t kInitialCapacity =
+      (GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD -
+       1) /
+      GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+
+  grpc_chttp2_hptbl() {
+    constexpr uint32_t AllocSize = sizeof(ents[0]) * kInitialCapacity;
+    grpc_mdelem* ents_new = static_cast<grpc_mdelem*>(gpr_malloc(AllocSize));
+    memset(ents_new, 0, AllocSize);
+    ents.reset(ents_new);
+  }
+
   /* the first used entry in ents */
-  uint32_t first_ent;
+  uint32_t first_ent = 0;
   /* how many entries are in the table */
-  uint32_t num_ents;
+  uint32_t num_ents = 0;
   /* the amount of memory used by the table, according to the hpack algorithm */
-  uint32_t mem_used;
+  uint32_t mem_used = 0;
   /* the max memory allowed to be used by the table, according to the hpack
      algorithm */
-  uint32_t max_bytes;
+  uint32_t max_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* the currently agreed size of the table, according to the hpack algorithm */
-  uint32_t current_table_bytes;
+  uint32_t current_table_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* Maximum number of entries we could possibly fit in the table, given defined
      overheads */
-  uint32_t max_entries;
+  uint32_t max_entries = kInitialCapacity;
   /* Number of entries allocated in ents */
-  uint32_t cap_entries;
+  uint32_t cap_entries = kInitialCapacity;
   /* a circular buffer of headers - this is stored in the opposite order to
      what hpack specifies, in order to simplify table management a little...
      meaning lookups need to SUBTRACT from the end position */
-  grpc_mdelem* ents;
-  grpc_mdelem static_ents[GRPC_CHTTP2_LAST_STATIC_ENTRY];
-} grpc_chttp2_hptbl;
+  grpc_core::UniqueArr<grpc_mdelem> ents;
```

Comment:
Instead of creating `grpc_core::UniqueArr<>`, how about just using `grpc_core::InlinedVector<>` here?

Reply:
I don't think InlinedVector is the right choice, for two reasons:
1) The dynamic resizing doesn't really help us since we're not always adding to the back; any resize implies a full rehash of all the table elements anyways.
2) Deciding what size to have inlined - one thought is to initialize the initial table size, but as soon as we have to grow past that, all those slots are going to be forever wasted. And to do a read, we'd have to always pay a branch penalty to see if we're hitting the inline buffer or not.

For that reason I'm either in favour of UniqueArr, *or* just raw pointers like before (with a preference for UniqueArr).

Reply:
Okay, that makes sense.

What I don't like about `UniqueArr<>` is that it leaves us with two different APIs, one for individual objects and another for arrays of objects.  In contast, `std::unique_ptr<>` handles both individual objects and arrays in the same interface, and our `UniquePtr<>` is supposed to be the equivalent of that.  If you can find a way to change `UniquePtr<>` to support the array case, that's fine.  Otherwise, let's just stick with the original code that uses raw pointers.

Reply:
Updated to used UniquePtr instead of UniqueArr. Removed UniqueArr.